pkg/scheduler/core/generic_scheduler.go

```diff
@@ -804,11 +804,9 @@ func PrioritizeNodes(
 		for j := range priorityConfigs {
 			result[i].Score += results[j][i].Score * priorityConfigs[j].Weight
 		}
-	}
 
-	for _, scoreList := range scoresMap {
-		for i := range nodes {
-			result[i].Score += scoreList[i]
+		for j := range scoresMap {
+			result[i].Score += scoresMap[j][i].Score
```

Comment:
I suspect that the original code will be faster because it has better cache locality when reading scores from the scoresMap.

Reply:
We need more evidence on this, before that I prefer to keep it this way.

Reply:
In theory sequential (or strided) reads are the fastest form of memory access patterns because of prefetching and cache locality. 

The other thing in this nested loop that could make it slower is that it makes n*m map lookups (n=number of nodes, m = number of priority functions/score plugins), while if we make the node loop inside, the number of scoresMap lookups will be zero. So it might be even faster (and easier to read) to flip the whole thing like this:

```
for i := range nodes {
		result = append(result, schedulerapi.HostPriority{Host: nodes[i].Name, Score: 0})
}

for j, priorityList := range priorityConfigs {
		for i := range nodes {
			result[i].Score += priorityList[i].Score * priorityConfigs[j].Weight
		}
}

for _, scoreList := range scoresMap {
		for i := range nodes {
			result[i].Score += scoreList[i].Score
		}
}
```

> We need more evidence on this, before that I prefer to keep it this way.
The theory says this change makes things slower, so if you don't have evidence to the contrary, then we shouldn't make the change, especially that it is not related to the purpose of the PR "using named array in normalize scores".

Reply:
> In theory sequential (or strided) reads are the fastest form of memory access patterns because of prefetching and cache locality.

I know the theroy of the cache locality

Here are the benchmark numbers:

```go
func locality(result []int, scores []int, anotherScores []int) {
	for i := range scores {
		result[i] = result[i] + scores[i]
	}

	for i := range anotherScores {
		result[i] = result[i] + anotherScores[i]
	}
}

func nonlocality(result []int, scores []int, anotherScores []int) {
	for i := range scores {
		result[i] = result[i] + scores[i]
		result[i] = result[i] + anotherScores[i]
	}
}

func BenchmarkLocality(b *testing.B) {
	result := make([]int, 5000)
	scores := make([]int, 5000)
	anotherScores := make([]int, 5000)

	for i := 0; i < 5000; i++ {
		scores[i] = i
		anotherScores[i] = i
	}

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		locality(result, scores, anotherScores)
	}
}

func BenchmarkNonLocality(b *testing.B) {
	result := make([]int, 5000)
	scores := make([]int, 5000)
	anotherScores := make([]int, 5000)

	for i := 0; i < 5000; i++ {
		scores[i] = i
		anotherScores[i] = i
	}

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		locality(result, scores, anotherScores)
	}
}

$ go test -bench=.
goos: darwin
goarch: amd64
pkg: github.com/golang
BenchmarkLocality-8      	  300000	      5049 ns/op
BenchmarkNonLocality-8   	  300000	      5041 ns/op
PASS
```

According to the benchmarks, I prefer to keep it this way.

Reply:
Strange, it is either golang's compiler is so good, or new architectures have such powerful memory controllers. I would like to do the same test in C++ to see if I get different results.

Do you have a theory why they are the same?

Reply:
The thing is that the number of priorities (`j` values) is very small, and so is the number of nodes. We are only talking about ~200kb, which can fit in the caches pretty easily.

Minor note in the posted benchmark: it should be `map[string][]int`. but again, we are only doing lookups in 20 values.

On the other hand, I don't understand the motivation to do the change. I'm not going to hold the review.

Reply:
I re ran the test and set the sizes to two orders of magnitude bigger, no difference.

The issue about cache is not only about size, but also what you move to the cache (how much of the cache line being moved from memory is actually used), so it is about effective memory bandwidth.